blockchain/thresholdstate.go

```diff
@@ -465,6 +465,91 @@ func (b *BlockChain) deploymentState(prevNode *blockNode, version uint32, deploy
 	return invalidState, DeploymentError(deploymentID)
 }
 
+// stateActivationNode returns the node at which the provided consensus
+// deployment agenda activated.
+func (b *BlockChain) stateActivationNode(version uint32, prevNode *blockNode, checker thresholdConditionChecker, cache *thresholdStateCache) (*blockNode, error) {
+	confirmationWindow := int64(checker.RuleChangeActivationInterval())
+	svh := checker.StakeValidationHeight()
+
+	// Return the threshold state for the window that contains the genesis
+	// block if the chain is not past stake validation height.
+	if prevNode == nil || prevNode.height+1 < svh+confirmationWindow {
+		return b.bestChain.Genesis(), nil
+	}
+
+	curState, err := b.thresholdState(version, prevNode, checker, cache)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the ancestor that is the first block of the previous confirmation
+	// window in order to get its threshold state.  This can be done because
+	// the state is the same for all blocks within a given window.
+	wantHeight := calcWantHeight(svh, confirmationWindow, prevNode.height+1)
+	prevNode = prevNode.Ancestor(wantHeight)
+	firstNodeLastConfWindow := *prevNode
+	for prevNode != nil {
+		state, err := b.thresholdState(version, prevNode, checker, cache)
+		if err != nil {
+			return nil, err
+		}
+
+		if state.State != curState.State {
+			stateActivationNode := firstNodeLastConfWindow.Ancestor(
+				prevNode.height + confirmationWindow + 1)
+			return stateActivationNode, nil
+		}
+
+		// Get the ancestor that is the first block of the previous
+		// confirmation window.
+		prevNode = prevNode.RelativeAncestor(confirmationWindow)
+	}
+
+	return b.bestChain.Genesis(), nil
+}
+
+// StateActivationHeight returns the height of the first block where the
+// current state of consensus deployment agenda activated.
+func (b *BlockChain) StateActivationHeight(hash *chainhash.Hash, version uint32, deploymentID string) (int64, error) {
```

Comment:
This is wrong too.  It's not returning when it activated, rather the height at which the state last changed.  For example, try calling this with the hash at height 140000.